### PR TITLE
A MethodHandles.Lookup API defineClass

### PR DESCRIPTION
**Pull request by ChengJin01**
The code is to introduce a new API called
defineClass which produces a class object
from the class bytes that shares the same
class loader, the same runtime package and
protection domain as the lookup class.

Signed-off-by: CHENGJin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Test Framework reviewer:  @renfeiw 
Peer reviewer:  @JasonFengJ9   (please double-check as the changes are reviewed previously)
Senior reviewer: @DanHeidinga

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -1629,21 +1632,59 @@ public VarHandle unreflectVarHandle(Field field) throws IllegalAccessException {
...
+				throw new NullPointerException(com.ibm.oti.util.Msg.getString("K065X")); //$NON-NLS-1$
+			}
+			
+			/* Extract the package name from the class bytes so as to check whether
+			 * the requested class shares the same package as the lookup class.
+			 */
+			ClassReader cr = new ClassReader(classBytes);
+			String targetClassName = cr.getClassName().replace('/', '.');
+			int separatorIndex = targetClassName.lastIndexOf('.');
+			String targetClassPackageName = null;
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: `targetClassPackageName` need to be initialized to an empty string instead cause `Class.getPackageName()` returns the empty string if the class is in an unnamed package. 
P.S. there is no need to initialize a variable to null.

Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Agreed and fixed.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -1629,21 +1632,59 @@ public VarHandle unreflectVarHandle(Field field) throws IllegalAccessException {
...
+				targetClassPackageName = targetClassName.substring(0, separatorIndex);
+			}
+			
+			String lookupClassPackageName = accessClass.getPackageName();
+			if (!Objects.equals(lookupClassPackageName, targetClassPackageName)) {
+				/*[MSG "K065Y", "The package of the requested class: {0} is different from the package of the lookup class: {1}"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065Y", targetClassPackageName, lookupClassPackageName)); //$NON-NLS-1$
+			}
+			
+			if (Lookup.PACKAGE != (Lookup.PACKAGE & accessMode)) {
```
Comment (JasonFengJ9) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: `PACKAGE` can be referred directly here.

Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Fixed as it is in the Lookup class.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import java.lang.reflect.Method;
+import java.lang.invoke.MethodHandles.Lookup;
+import static java.lang.invoke.MethodHandles.lookup;
+import com.ibm.j9.jsr292.ClassBytes;
+
+public class MethodHandleAPI_defineClass {
+	final static Lookup lookup = lookup();
+	
+	/**
+	 * defineClass test for a null array of class types
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: is this 'class bytes' instead?

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	 */
+	@Test(expectedExceptions = NullPointerException.class, groups = { "level.sanity" })
+	public static void test_defineClass_ClassBytes_NullArray() throws Throwable {
+		Class<?> targetClass = lookup.defineClass(null);
+		Assert.fail("The test case failed to detect a null array of class types");
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: same here.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	 */
+	@Test(expectedExceptions = IllegalAccessException.class, groups = { "level.sanity" })
+	public static void test_defineClass_Lookup_WithoutPackageMode() throws Throwable {
+		Lookup newLookup = lookup.dropLookupMode(Lookup.PACKAGE);
+		Class<?> targetClass = newLookup.defineClass(ClassBytes.validClassBytes_SamePackage());
+		Assert.fail("The test case failed to detect that the lookup doesn't have the PACKAGE access mode");
+	}
+	
+	/**
+	 * defineClass test for an corrupted class byte array
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: it should be 'a'.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	 */
+	@Test(expectedExceptions = LinkageError.class, groups = { "level.sanity" })
+	public static void test_defineClass_ClassBytes_CorruptedArray() throws Throwable {
+		Class<?> targetClass = lookup.defineClass(ClassBytes.invalidClassBytes_SamePackage());
+		Assert.fail("The test case failed to detect an corrupted class byte array");
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: same here.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	}
+	
+	/**
+	 * defineClass test for a valid class byte array that shares the same class loader
+	 * , the same runtime package and the protection domain as the lookup class.
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: Formatting nit: the comma is better after 'loader'.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_privateLookupIn.java [hunk 8]
```diff
@@ -109,8 +109,8 @@ public static void test_privateLookupIn_TargetClass_Itself_UnnamedModule() throw
 		Lookup lookup = MethodHandles.lookup();
 		Class<?> targetClass = lookup.lookupClass();
 		Lookup mhprivateLookupIn = privateLookupIn(targetClass, lookup);
-		Assert.assertEquals(targetClass, mhprivateLookupIn.lookupClass());
-		Assert.assertEquals(Lookup.PRIVATE, mhprivateLookupIn.lookupModes() & Lookup.PRIVATE);
+		Assert.assertEquals(mhprivateLookupIn.lookupClass(), targetClass);
+		Assert.assertEquals(mhprivateLookupIn.lookupModes() & Lookup.PRIVATE, Lookup.PRIVATE);
```
Comment (JasonFengJ9) [inline test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_privateLookupIn.java]: Any particular reason to switch the order here? (assuming this is not related to `MH_defineClass` changes.)

Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_privateLookupIn.java]: Assert.assertEquals() in testNG is different from Junit in terms of signature as follows:
Assert.assertEquals(acutal_result, expected_result)

(please take reference to https://jitpack.io/com/github/cbeust/testng/master-6.12-g62083d8-127/javadoc/org/testng/Assert.html if interested)

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java (hunk 3 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: Updated.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java (hunk 4 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: Updated.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java (hunk 5 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: Updated.

File: test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java (hunk 7 above)
Comment (ChengJin01) [reply test/OpenJ9_Jsr_292_API/src/com/ibm/j9/jsr292/api/MethodHandleAPI_defineClass.java]: Updated.

**Comment by JasonFengJ9** [comment]
lgtm

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 9]
```diff
@@ -1629,21 +1632,67 @@ public VarHandle unreflectVarHandle(Field field) throws IllegalAccessException {
...
+			/* Extract the package name from the class bytes so as to check whether
+			 * the requested class shares the same package as the lookup class.
+			 * Note: the class bytes should be considered as corrupted if any exception
+			 *       is captured in ASM class reader.
+			 */
+			ClassReader cr = null;
+			try {
+				cr = new ClassReader(classBytes);
+			} catch (ArrayIndexOutOfBoundsException e) {
+				throw new ClassFormatError(com.ibm.oti.util.Msg.getString("K065Y1")); //$NON-NLS-1$
```
Comment (pshipton) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: The external message K065Y1 needs to be declared.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 10]
```diff
@@ -1629,21 +1632,67 @@ public VarHandle unreflectVarHandle(Field field) throws IllegalAccessException {
...
+				targetClassPackageName = targetClassName.substring(0, separatorIndex);
+			}
+			
+			String lookupClassPackageName = accessClass.getPackageName();
+			if (!Objects.equals(lookupClassPackageName, targetClassPackageName)) {
+				/*[MSG "K065Y2", "The package of the requested class: {0} is different from the package of the lookup class: {1}"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065Y2", targetClassPackageName, lookupClassPackageName)); //$NON-NLS-1$
+			}
+			
+			if (PACKAGE != (PACKAGE & accessMode)) {
```
Comment (pshipton) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: It would make sense to do this check earlier as if it fails using the ClassReader is not necessary.
Similarly for the SecurityManager check below.

**Comment by pshipton** [comment]
Does the implementation pass the JCK test?
api/java_lang/invoke/MethodHandles/Lookup/index.html#DefineClass

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 9 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Already updated.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 10 above)
Comment (ChengJin01) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Agreed and updated as suggestion above.

**Comment by ChengJin01** [comment]
The code was verified/passed on the latest JCK tests (b68) with api/java_lang/invoke/MethodHandles/Lookup/index.html#DefineClass

**Comment by pshipton** [comment]
@llxia any concerns with the tests?

**Comment by pshipton** [comment]
jenkins test sanity plinux

**Comment by pshipton** [comment]
@llxia any concerns with the testing updates?

**Comment by llxia** [comment]
@sophia-guo , could you please review the test changes? Thanks.

**Comment by ChengJin01** [comment]
Test cases updated as suggested above.

**Comment by pshipton** [comment]
jenkins test sanity plinux
